app/controllers/incomeSources/manage/ConfirmReportingMethodSharedController.scala

```diff
@@ -59,24 +62,50 @@ class ConfirmReportingMethodSharedController @Inject()(val manageIncomeSources:
                                                        implicit val appConfig: FrontendAppConfig) extends ClientConfirmedController
   with FeatureSwitching with IncomeSourcesUtils {
 
-  def show(id: Option[String],
-           taxYear: String,
+  def show(taxYear: String,
            changeTo: String,
            isAgent: Boolean,
            incomeSourceType: IncomeSourceType
           ): Action[AnyContent] = authenticatedAction(isAgent) { implicit user =>
 
-    handleShowRequest(taxYear, changeTo, isAgent, incomeSourceType, id)
+    withIncomeSourcesFS {
+      sessionService.get(SessionKeys.incomeSourceId).flatMap {
+        case Right(incomeSourceIdMayBe) =>
+          incomeSourceIdMayBe match {
+            case Some(incomeSourceId) => handleShowRequest(taxYear, changeTo, isAgent, incomeSourceType, Some(incomeSourceId))
+            case None => Future.failed(MissingSessionKey(incomeSourceId))
+          }
+        case Left(exception) => Future.failed(exception)
+      }
+    }.recover {
+      case exception =>
+        val errorHandler = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+        Logger("application").error(s"[ConfirmReportingMethodSharedController][show] ${exception.getMessage}")
+        errorHandler.showInternalServerError()
+    }
   }
 
-  def submit(id: String,
-             taxYear: String,
+  def submit(taxYear: String,
              changeTo: String,
              isAgent: Boolean,
              incomeSourceType: IncomeSourceType
             ): Action[AnyContent] = authenticatedAction(isAgent) { implicit user =>
 
-    handleSubmitRequest(taxYear, changeTo, isAgent, id, incomeSourceType)
+    withIncomeSourcesFS {
+      sessionService.get(SessionKeys.incomeSourceId).flatMap {
+        case Right(incomeSourceIdMayBe) =>
+          incomeSourceIdMayBe match {
```

Comment:
you can change handleSubmitRequest signature to permit incomeSourceIdMayBe ( as Optional[String]), in order to reduce this code to next one:

```
    withIncomeSourcesFS {
      sessionService.get(SessionKeys.incomeSourceId).flatMap {
        case Right(incomeSourceIdMayBe) =>
          handleSubmitRequest(taxYear, changeTo, isAgent, incomeSourceId, incomeSourceIdMayBe)
        case Left(exception) => 
          Future.failed(exception)
      }
    }.recover {
```